In [356]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings

warnings.filterwarnings('ignore')

In [358]:
from sklearn.datasets import make_classification

In [360]:
#create dataset 
X,y = make_classification(n_samples=100,n_features=10,n_classes=2,random_state=42)

In [362]:
X

array([[-1.14052601,  1.35970566,  0.86199147,  0.84609208,  0.60600995,
        -1.55662917,  1.75479418,  1.69645637, -1.28042935, -2.08192941],
       [-0.07873421, -1.32933233,  0.6273745 , -1.19300559, -0.77300978,
         0.0976761 ,  0.49799829,  0.95927083,  0.02451017,  1.45114361],
       [ 0.80742726,  0.73019848, -1.28568005,  0.88948365, -1.8048821 ,
        -0.76325916,  0.04808495, -0.90431663, -1.62754244,  0.2597225 ],
       [ 0.58846525, -0.3751207 , -0.57500215, -0.14951801, -0.56372455,
         0.41293145,  0.24368721, -0.50694318, -0.8222204 ,  0.24496657],
       [ 1.63631177, -1.64060704, -1.36045573, -0.94116325, -1.43014138,
         1.6324113 ,  0.13074058, -1.43586215, -0.44004449,  1.44127329],
       [-0.47936995, -1.08324727,  1.02255619, -1.09939128, -2.30192116,
        -0.2750517 ,  1.36687427, -0.24903604, -1.51519106,  1.64496771],
       [ 2.48904785, -2.52343407, -2.05832072, -1.45612516, -0.05952536,
        -0.92693047, -1.02438764, -1.24778318

In [364]:
y

array([1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0])

In [366]:
pd.DataFrame(X)

0         1         2         3         4         5         6  \
0  -1.140526  1.359706  0.861991  0.846092  0.606010 -1.556629  1.754794   
1  -0.078734 -1.329332  0.627375 -1.193006 -0.773010  0.097676  0.497998   
2   0.807427  0.730198 -1.285680  0.889484 -1.804882 -0.763259  0.048085   
3   0.588465 -0.375121 -0.575002 -0.149518 -0.563725  0.412931  0.243687   
4   1.636312 -1.640607 -1.360456 -0.941163 -1.430141  1.632411  0.130741   
..       ...       ...       ...       ...       ...       ...       ...   
95  0.531047  0.948005 -1.032233  0.996309 -1.008086 -2.038125 -0.351513   
96 -0.963142  0.556003  0.964233  0.193754 -0.703176  1.149273  1.770801   
97  0.950573 -2.255540 -0.270624 -1.692005  0.504047 -0.985726 -0.792873   
98 -1.016951  0.554484  1.031102  0.175930  0.615936 -1.081063 -0.309546   
99  1.897924 -1.614533 -1.693028 -0.838070  0.791663  1.158111  0.628346   

           7         8         9  
0   1.696456 -1.280429 -2.081929  
1   0.959271  0.024510  1.451144  
2  -0.904317 -1.627542  0.259723  
3  -0.506943 -0.822220  0.244967  
4  -1.435862 -0.440044  1.441273  
..       ...       ...       ...  
95  1.676437 -1.870792  0.018418  
96  1.812449 -0.034988 -0.626967  
97 -1.035242 -0.530258 -0.107030  
98 -1.251114  0.593101  0.326133  
99 -0.897254  0.624120 -0.012247  

[100 rows x 10 columns]

In [368]:
y

array([1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0])

In [370]:
pd.DataFrame(y)

0
0   1
1   1
2   0
3   0
4   0
.. ..
95  0
96  1
97  1
98  1
99  0

[100 rows x 1 columns]

In [372]:
# train test split 
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.30,random_state=42)

In [374]:
# Model training 
from sklearn.linear_model import LogisticRegression
logistic = LogisticRegression()

In [376]:
logistic.fit(X_train,y_train)

LogisticRegression()

In [378]:
y_test

array([0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 1, 1, 0, 1])

In [380]:
y_pred = logistic.predict(X_test)
print(y_pred)

[0 1 1 1 1 0 0 1 1 1 0 1 1 0 1 0 1 0 0 0 1 0 0 0 1 1 1 1 1 1]


In [382]:
logistic.predict_proba(X_test)  # refer both notes 

array([[7.49632742e-01, 2.50367258e-01],
       [5.71492961e-02, 9.42850704e-01],
       [8.47383467e-03, 9.91526165e-01],
       [3.84386602e-02, 9.61561340e-01],
       [2.79853023e-02, 9.72014698e-01],
       [9.64640318e-01, 3.53596817e-02],
       [9.65369374e-01, 3.46306260e-02],
       [2.13024820e-01, 7.86975180e-01],
       [2.27546362e-01, 7.72453638e-01],
       [1.81401566e-01, 8.18598434e-01],
       [9.87298049e-01, 1.27019510e-02],
       [1.84390109e-03, 9.98156099e-01],
       [1.11638160e-03, 9.98883618e-01],
       [9.98846835e-01, 1.15316489e-03],
       [3.19528899e-03, 9.96804711e-01],
       [9.75335190e-01, 2.46648100e-02],
       [2.79003416e-02, 9.72099658e-01],
       [9.72840626e-01, 2.71593739e-02],
       [9.57717728e-01, 4.22822718e-02],
       [9.96764165e-01, 3.23583540e-03],
       [3.26255970e-02, 9.67374403e-01],
       [9.91543744e-01, 8.45625612e-03],
       [6.48321099e-01, 3.51678901e-01],
       [9.99581102e-01, 4.18898067e-04],
       [4.128605

onnamathe test data yude 0 classify leek avaanulla probabilty 0.7496 ie 7.496 , athpole 1 classify leek avanulla 
probababilty 0.2503 ie 2.503 aahn so, onnamatthhe test data yude data , 0 classify category yilaaahn

In [385]:
#performance metrics 

In [387]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [389]:
score = accuracy_score(y_test,y_pred)
print(score)
cm = confusion_matrix(y_test,y_pred)
print(cm)
print(classification_report(y_test,y_pred))

0.9333333333333333
[[12  2]
 [ 0 16]]
              precision    recall  f1-score   support

           0       1.00      0.86      0.92        14
           1       0.89      1.00      0.94        16

    accuracy                           0.93        30
   macro avg       0.94      0.93      0.93        30
weighted avg       0.94      0.93      0.93        30



In [391]:
logistic.intercept_

array([0.24283506])

In [393]:
logistic.coef_

array([[-1.43936447, -0.58123104,  2.00445388, -0.95214288,  0.29119542,
         0.02499908, -0.04863764, -0.0487132 ,  0.53322987, -0.09783875]])

### Hyperparameter Tuning and cross validation

## Grid Search Cv 

In [397]:
model = LogisticRegression()
penalty = ['l1','l2','elasticnet']
c_values = [100,10,1.0,0.1,0.01]
solver = ['newton-cg','lbfgs','liblinear','sag','saga']

In [399]:
params = dict(penalty=penalty,C=c_values,solver=solver)

In [401]:
from sklearn.model_selection import StratifiedKFold
cv = StratifiedKFold()

In [403]:
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(estimator=model,param_grid=params,scoring='accuracy',cv=cv,n_jobs=-1)

In [405]:
grid

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': [100, 10, 1.0, 0.1, 0.01],
                         'penalty': ['l1', 'l2', 'elasticnet'],
                         'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag',
                                    'saga']},
             scoring='accuracy')

In [407]:
grid.fit(X_train,y_train)

/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linea

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': [100, 10, 1.0, 0.1, 0.01],
                         'penalty': ['l1', 'l2', 'elasticnet'],
                         'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag',
                                    'saga']},
             scoring='accuracy')

In [408]:
grid.best_params_

{'C': 100, 'penalty': 'l1', 'solver': 'liblinear'}

In [409]:
grid.best_score_

0.9857142857142858

In [413]:
y_pred = grid.predict(X_test)

In [415]:
score = accuracy_score(y_test,y_pred)
print(score)
cm = confusion_matrix(y_test,y_pred)
print(cm)
print(classification_report(y_test,y_pred))

0.9
[[11  3]
 [ 0 16]]
              precision    recall  f1-score   support

           0       1.00      0.79      0.88        14
           1       0.84      1.00      0.91        16

    accuracy                           0.90        30
   macro avg       0.92      0.89      0.90        30
weighted avg       0.92      0.90      0.90        30



## Randomized Search Cv

In [418]:
model = LogisticRegression()
from sklearn.model_selection import RandomizedSearchCV
randomcv = RandomizedSearchCV(estimator = model, param_distributions = params , scoring = 'accuracy', cv = 5)

In [420]:
randomcv

RandomizedSearchCV(cv=5, estimator=LogisticRegression(),
                   param_distributions={'C': [100, 10, 1.0, 0.1, 0.01],
                                        'penalty': ['l1', 'l2', 'elasticnet'],
                                        'solver': ['newton-cg', 'lbfgs',
                                                   'liblinear', 'sag',
                                                   'saga']},
                   scoring='accuracy')

In [422]:
randomcv.fit(X_train,y_train)

RandomizedSearchCV(cv=5, estimator=LogisticRegression(),
                   param_distributions={'C': [100, 10, 1.0, 0.1, 0.01],
                                        'penalty': ['l1', 'l2', 'elasticnet'],
                                        'solver': ['newton-cg', 'lbfgs',
                                                   'liblinear', 'sag',
                                                   'saga']},
                   scoring='accuracy')

In [424]:
randomcv.best_score_

0.9857142857142858

In [426]:
randomcv.best_params_

{'solver': 'liblinear', 'penalty': 'l1', 'C': 1.0}

In [428]:
y_predr  = randomcv.predict(X_test)

In [430]:
score = accuracy_score(y_test,y_predr)
print(score)
cm = confusion_matrix(y_test,y_predr)
print(cm)
print(classification_report(y_test,y_predr))

0.9666666666666667
[[13  1]
 [ 0 16]]
              precision    recall  f1-score   support

           0       1.00      0.93      0.96        14
           1       0.94      1.00      0.97        16

    accuracy                           0.97        30
   macro avg       0.97      0.96      0.97        30
weighted avg       0.97      0.97      0.97        30



# Logistic Regression For MultiClass classification Problem 

In [433]:
X,y = make_classification(n_samples=100,n_features=10,n_informative=3,n_classes=3,random_state=42)

In [435]:
y

array([0, 1, 0, 1, 1, 0, 2, 1, 1, 2, 0, 0, 0, 2, 1, 0, 2, 2, 2, 1, 0, 0,
       1, 1, 2, 2, 2, 1, 2, 2, 0, 0, 2, 2, 2, 0, 2, 2, 2, 0, 2, 0, 2, 0,
       1, 0, 2, 0, 1, 0, 1, 1, 2, 2, 1, 1, 2, 0, 0, 0, 0, 1, 2, 1, 0, 1,
       1, 0, 0, 1, 1, 2, 0, 2, 2, 1, 1, 0, 2, 0, 1, 0, 2, 2, 1, 0, 1, 2,
       1, 1, 0, 1, 0, 2, 0, 2, 0, 2, 1, 1])

In [437]:
pd.DataFrame(y)

0
0   0
1   1
2   0
3   1
4   1
.. ..
95  2
96  0
97  2
98  1
99  1

[100 rows x 1 columns]

In [439]:
# train test split 
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.30,random_state=42)

In [441]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression(multi_class='ovr')
logreg.fit(X_train,y_train)

LogisticRegression(multi_class='ovr')

In [443]:
y_pred = logreg.predict(X_test)
y_pred

array([2, 2, 0, 0, 2, 0, 1, 0, 0, 1, 2, 2, 2, 2, 0, 0, 2, 1, 1, 1, 2, 0,
       1, 2, 0, 0, 1, 1, 2, 0])

In [445]:
logreg.coef_

array([[-0.44106696, -0.08708238,  0.55557675,  0.23558516,  0.32803244,
        -0.24877277,  0.00748971,  0.87552196, -0.38801939,  0.42601779],
       [-0.07104312,  0.43251239,  0.10665511, -0.75440983,  0.09581888,
         0.23125788,  0.08146246,  0.23764653,  0.5537666 , -0.73054901],
       [ 0.45626522, -0.21515342, -0.53028333,  0.67919173, -0.37035808,
         0.12599741, -0.19539769, -1.08016074, -0.36905125,  0.36672747]])

In [447]:
logreg.intercept_

array([-0.75351502, -0.70077734, -1.66081075])

In [449]:
score = accuracy_score(y_test,y_pred)
print(score)
cm = confusion_matrix(y_test,y_pred)
print(cm)
print(classification_report(y_test,y_pred))

0.4666666666666667
[[6 5 1]
 [5 1 3]
 [0 2 7]]
              precision    recall  f1-score   support

           0       0.55      0.50      0.52        12
           1       0.12      0.11      0.12         9
           2       0.64      0.78      0.70         9

    accuracy                           0.47        30
   macro avg       0.44      0.46      0.45        30
weighted avg       0.45      0.47      0.45        30



model varacha line il prashnmund hyperparameter cheythintt sheriyakanm 

### hyperparameter tuning 

In [501]:
model1 = LogisticRegression()
penalty = ['l1','l2','elasticnet']
c = [0.001, 0.01, 0.1, 1, 10]
solver = ['lbfgs', 'liblinear', 'newton-cg', 'newton-cholesky', 'sag', 'saga']
iter = [100,200,300,350]
multiclass = ['ovr','multinomial']
parameter = dict(penalty=penalty,C=c,solver=solver,max_iter=iter,multi_class=multiclass)

In [503]:
parameter

{'penalty': ['l1', 'l2', 'elasticnet'],
 'C': [0.001, 0.01, 0.1, 1, 10],
 'solver': ['lbfgs',
  'liblinear',
  'newton-cg',
  'newton-cholesky',
  'sag',
  'saga'],
 'max_iter': [100, 200, 300, 350],
 'multi_class': ['ovr', 'multinomial']}

In [505]:
from sklearn.model_selection import StratifiedKFold
cv = StratifiedKFold()
from sklearn.model_selection import GridSearchCV
grid2 = GridSearchCV(estimator=model1,param_grid=parameter,scoring='accuracy',cv=cv,n_jobs=-1)

In [507]:
grid2

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': [0.001, 0.01, 0.1, 1, 10],
                         'max_iter': [100, 200, 300, 350],
                         'multi_class': ['ovr', 'multinomial'],
                         'penalty': ['l1', 'l2', 'elasticnet'],
                         'solver': ['lbfgs', 'liblinear', 'newton-cg',
                                    'newton-cholesky', 'sag', 'saga']},
             scoring='accuracy')

In [509]:
grid2.fit(X_train,y_train)

/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/home/fadhil/anaconda3/lib/python3.12/site-packages/sklearn/linea

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': [0.001, 0.01, 0.1, 1, 10],
                         'max_iter': [100, 200, 300, 350],
                         'multi_class': ['ovr', 'multinomial'],
                         'penalty': ['l1', 'l2', 'elasticnet'],
                         'solver': ['lbfgs', 'liblinear', 'newton-cg',
                                    'newton-cholesky', 'sag', 'saga']},
             scoring='accuracy')

In [510]:
grid2.best_score_

0.5714285714285714

In [513]:
grid2.best_params_

{'C': 0.1,
 'max_iter': 100,
 'multi_class': 'ovr',
 'penalty': 'l2',
 'solver': 'lbfgs'}

In [515]:
y_predict = grid2.predict(X_test)

In [517]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [519]:
score = accuracy_score(y_test,y_predict)
cm = confusion_matrix(y_test,y_predict)
print(score)
print(cm)
print(classification_report(y_test,y_predict))

0.4666666666666667
[[6 5 1]
 [5 1 3]
 [0 2 7]]
              precision    recall  f1-score   support

           0       0.55      0.50      0.52        12
           1       0.12      0.11      0.12         9
           2       0.64      0.78      0.70         9

    accuracy                           0.47        30
   macro avg       0.44      0.46      0.45        30
weighted avg       0.45      0.47      0.45        30



In [521]:
grid2.best_estimator_

LogisticRegression(C=0.1, multi_class='ovr')